src/main/java/org/truffleruby/core/regexp/TruffleRegexpNodes.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) 2017 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 1.0
+ * GNU General Public License version 2
+ * GNU Lesser General Public License version 2.1
+ */
+package org.truffleruby.core.regexp;
+
+import org.joni.Regex;
+import org.truffleruby.Layouts;
+import org.truffleruby.builtins.CoreClass;
+import org.truffleruby.builtins.CoreMethod;
+import org.truffleruby.builtins.CoreMethodArrayArgumentsNode;
+import org.truffleruby.core.regexp.RegexpNodes.ToSNode;
+import org.truffleruby.core.rope.CodeRange;
+import org.truffleruby.core.rope.Rope;
+import org.truffleruby.core.string.StringNodes.StringAppendPrimitiveNode;
+import org.truffleruby.core.string.StringOperations;
+
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.ExplodeLoop;
+import com.oracle.truffle.api.object.DynamicObject;
+import com.oracle.truffle.api.object.DynamicObjectFactory;
+
+@CoreClass("Truffle::RegexpOperations")
+public class TruffleRegexpNodes {
+
+    @CoreMethod(names = "union", onSingleton = true, required = 2, rest = true)
+    public static abstract class RegexpUnionNode extends CoreMethodArrayArgumentsNode {
+
+        @Child StringAppendPrimitiveNode appendNode = StringAppendPrimitiveNode.create();
+        @Child ToSNode toSNode = ToSNode.create();
+
+        @Specialization(guards = "argsMatch(cachedArgs, args)")
+        public DynamicObject executeFastUnion(VirtualFrame frame, DynamicObject str, DynamicObject sep, Object[] args,
+                @Cached(value = "args", dimensions = 1) Object[] cachedArgs,
+                @Cached("buildUnion(frame, str, sep, args)") DynamicObject union) {
+            return union;
+        }
```

Comment:
Caching the Regexp instance is user-visible.
I think Regexp are rarely used to store state, but something to be aware of.
Could we cache the Joni Regex but not the Ruby Regexp object to respect semantics?

Reply:
We'd probably want to share the encoding cache and so forth as well, and revisit the other `Regexp#to_s` and interpolated regexp changes to check the internals of the regexp object rather than the object identity itself.